torch/utils/flop_counter.py

```diff
@@ -0,0 +1,325 @@
+        return (x,)
+    return x
+
+
+# Thanks BingChat!
+def convert_num_to_suffix_str(number):
+    # Define the suffixes for different orders of magnitude
+    suffixes = ["", "K", "M", "B", "T"]
+    # Find the index of the appropriate suffix based on the number of digits
+    index = max(0, min(len(suffixes) - 1, (len(str(number)) - 1) // 3))
+    # Divide the number by 1000^index and format it to two decimal places
+    value = "{:.2f}".format(number / (1000 ** index))
+    # Return the value and the suffix as a string
+    return value + suffixes[index]
+
+class FlopCounterMode(TorchDispatchMode):
+    """
+    ``FlopCounterMode`` is a context manager that counts the number of
+    flops within its context. It does this using a ``TorchDispatchMode``.
+
+    It also supports hierarchical output by passing a module (or list of modules) to FlopCounterMode on construction.
+
+    Example usage
+
+    .. code-block:: python
+
+        mod = ...
+        flop_counter = FlopCounterMode(mod)
+        with flop_counter:
+            mod.sum().backward()
+
+    """
+    def __init__(
+            self,
+            mods: Optional[Union[torch.nn.Module, List[torch.nn.Module]]] = None,
+            depth: int = 2,
+            display: bool = True,
+            custom_mapping: Dict[Any, Any] = None):
+        self.flop_counts: Dict[str, Dict[Any, int]] = defaultdict(lambda: defaultdict(int))
+        self.depth = depth
+        self.parents = ["Global"]
+        self.display = display
+        if custom_mapping is None:
+            custom_mapping = {}
+        if isinstance(mods, torch.nn.Module):
+            mods = [mods]
+        self.mods = mods
+        if mods is not None:
+            for mod in mods:
+                prefix = type(mod).__name__
+                for name, module in dict(mod.named_modules()).items():
+                    if name == "":
+                        name = prefix
+                    else:
+                        name = ".".join([prefix, name])
+                    module.register_forward_pre_hook(self._enter_module(name))
+                    module.register_forward_hook(self._exit_module(name))
+        self.flop_mapping = {**flop_mapping, **custom_mapping}
+
+    def _enter_module(self, name):
+        def f(module, inputs):
+            inputs = normalize_tuple(inputs)
+            out = self._create_pre_module(name)(*inputs)
+            return out
+
+        return f
+
+    def _exit_module(self, name):
+        def f(module, inputs, outputs):
+            outputs = normalize_tuple(outputs)
+            return self._create_post_module(name)(*outputs)
+        return f
+
+    def _create_post_module(self, name):
+        class PushState(torch.autograd.Function):
+            @staticmethod
+            def forward(ctx, *args):
+                assert(self.parents[-1] == name)
+                self.parents.pop()
+                args = tree_map(lambda x: x.clone() if isinstance(x, torch.Tensor) else x, args)
```

Comment:
quick question: why are we calling `clone()` here?

Reply:
If the output of an autograd.Function is a view of the input, there are some restrictions on whether you can use inplace operations or not.